File: Leetcode/Python/_289.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        neighbours = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
        
        row_len = len(board)
        col_len = len(board[0])
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                
                live_neighbours = 0
                for neighbour in neighbours:
                    
                    row = i + neighbour[0]
                    col = j + neighbour[1]
                    
                    if (row > -1 and row < row_len) and (col > -1 and col < col_len) and abs(board[row][col]) == 1:
                        live_neighbours += 1

                if (live_neighbours < 2 or live_neighbours > 3) and abs(board[i][j]) == 1:
                    board[i][j] = -1 # was alive -> now dead

                    
                elif live_neighbours == 3 and board[i][j] == 0:
                    board[i][j] = 2 # was dead -> now alive

        for i in range(len(board)):
            for j in range(len(board[0])):
                board[i][j] = 0 if board[i][j] == 0 or board[i][j] == -1 else 1
        

File: Leetcode/cpp/289.cpp
class Solution {
private:
    vector<vector<int>> dirs {{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {-1, 1}, {1, -1}};

public:
    int getNeighbours(vector<vector<int>>& board, int row, int col, int maxRows, int maxCols) {
        int alive = 0;
        for (int i = 0; i < dirs.size(); ++i) {
            int newRow = row + dirs[i][0];
            int newCol = col + dirs[i][1];
            if (newRow >= 0 && newRow < maxRows && newCol >= 0 && newCol < maxCols) {
                alive += board[newRow][newCol];
            }
        }
        return alive;
    }
    void gameOfLife(vector<vector<int>>& board) {
        std::set<pair<int, int>> alive, dead;
        int maxRows = board.size();
        int maxCols = board[0].size();
        for (int i = 0; i < maxRows; ++i) {
            for (int j = 0; j < maxCols; ++j) {
                int nei = getNeighbours(board, i, j, maxRows, maxCols);
                if (board[i][j] == 0 && nei == 3) {
                    alive.insert({i, j});
                } else if (board[i][j] && (nei > 3 || nei < 2)) {
                    dead.insert({i, j});
                }
            }
        }

        for (int i = 0; i < maxRows; ++i) {
            for (int j = 0; j < maxCols; ++j) {
                std::pair<int, int> cell = {i, j};
                if (dead.contains(cell)) {
                    board[i][j] = 0;
                } else if (alive.contains(cell)) {
                    board[i][j] = 1;
                }
            }
        }
    }
};
